### ISSUE TITLE
Incorrect Cholesky backward grad due to incorrect symmetrization

### ISSUE THREAD

**Issue by ayghri** [pytorch/pytorch#137284]
### 🐛 Describe the bug

I believe `cholesky_backward` implementation is incorrect due to the false but well-intentioned assumption that $\overline{\Sigma}=\frac{S + S^\top}{2}$ is the backward gradient for $\Sigma$ once we get the backward gradient $S$, but the math tells us that if $dy = Tr(Sd\Sigma)$ then  $\overline{\Sigma}=S + S^\top -diag(S)$.

In the current implementation, the former incorrect approach is used. 
We can prove that $S = L^{-\top}\Phi(L^\top \overline{L})L^{-1}$ (See [paper](https://arxiv.org/abs/1602.07527)) where $\Phi(B)=tril(B) - \frac{1}{2}diag(B)$ (take the lower triangular and halves the diagonal).

## Current implementation
To make sense of the current implementation, it uses
$\overline{\Sigma}= L^{-\top}\left[\frac{\Phi(L^\top \overline{L}) + \Phi(L^\top \overline{L})^\top}{2}\right]L^{-1}$
Using $M = L^\top \overline{L}$:
$$\frac{\Phi(L^\top \overline{L}) + \Phi(L^\top \overline{L})^\top}{2}=\frac{1}{2}\left[tril(M) + tril(M)^\top -diag(M)\right]$$
with $A=tril(M)$ it's
$$\frac{1}{2}\left[A + A^\top -diag(A)\right] =\left[A + tril(A,-1)^\top\right]$$ (always true if A is lower triangular)
## The math way
Compute $S$ from above and return $S + S^\top -diag(S)$. I couldn't come up with a more concise way using the tril(A,-1) because of diag(S).

## Examples
Let's see some numbers.

Consider the symmetric PD matrix with $s,u>0$ and $u>\frac{t^2}{s}$
$$\Sigma = [[s , t],  [t , u]]$$

Cholesky decomposition is:
$$L = [[\sqrt{s} , 0 ],[\frac{t}{\sqrt{s}} , \sqrt{u - \frac{t^2}{s}}]]$$

Let's take s,t,u = 1,1,2 and backwards $L_{1,2}$ , we expect $\overline{t}=\frac{1}{\sqrt(s)}=1$.
What we get:
```python
import torch
## Current result
S = torch.tensor([[1.,1.],[1.,2.]], requires_grad=True)
A = torch.tril(torch.ones((2,2)),-1)
# >> A
# tensor([[0., 0.],
#         [1., 0.]])
L = torch.linalg.cholesky(S)
loss = (L*A).sum()
loss.backward(reta)
# >> S.grad[0,1]
# tensor(0.5000)

## What we should get if we manually compute L =cholesky(S)
S = torch.tensor([[1.,1.],[1.,2.]], requires_grad=True)
manual_L = torch.stack([torch.cat([S[0][:1]**.5, torch.tensor([0.0])]),
    torch.cat([S[0][1:]/S[0][:1]**.5, (S[1][1:] - S[0][1:]**2/S[0][:1])**.5])])
true_loss = torch.sum(e_L*A)
true_loss.backward()
# >> S.grad[0,1]
# tensor(1.)
```

```python
import torch
torch.manual_seed(2024)
C=torch.randn(4,4)/4.0
S = C.t().mm(C)
S.requires_grad_()
A = torch.randn(4,4)/4.0
loss = (A*torch.linalg.cholesky(S)).sum()
loss.backward()
S.grad
## current grad
tensor([[4.8560, 5.0956, 4.2147, 2.6087],
        [5.0956, 4.6948, 3.8186, 2.2791],
        [4.2147, 3.8186, 3.4634, 1.8416],
        [2.6087, 2.2791, 1.8416, 0.7801]])

## proposed pull request
tensor([[ 4.8560, 10.1912,  8.4293,  5.2173],
        [10.1912,  4.6948,  7.6372,  4.5582],
        [ 8.4293,  7.6372,  3.4634,  3.6832],
        [ 5.2173,  4.5582,  3.6832,  0.7801]])
```

## TLDR
Incorrect `cholesky_backward` due to incorrect symmetrization of the backward sensitivity using `(S + S^H)/2` which halves the non-diagonal elements. The math tells us that we should mirror the non-diagonal sensitivities while keeping the diagonal ones using `S + S^H - diag(S)`.

### Versions

Collecting environment information...
PyTorch version: 2.4.1+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Gentoo Linux (x86_64)
GCC version: (Gentoo 13.3.1_p20240614 p17) 13.3.1 20240614
Clang version: 18.1.8
CMake version: version 3.28.5
Libc version: glibc-2.39

Python version: 3.12.3 (main, Jun 11 2024, 01:10:37) [GCC 13.2.1 20240210] (64-bit runtime)
Python platform: Linux-6.6.38-gentoo-x86_64-x86_64-11th_Gen_Intel-R-_Core-TM-_i7-11700K_@_3.60GHz-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3090
Nvidia driver version: 550.107.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        39 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               16
On-line CPU(s) list:                  0-15
Vendor ID:                            GenuineIntel
Model name:                           11th Gen Intel(R) Core(TM) i7-11700K @ 3.60GHz
CPU family:                           6
Model:                                167
Thread(s) per core:                   2
Core(s) per socket:                   8
Socket(s):                            1
Stepping:                             1
CPU(s) scaling MHz:                   81%
CPU max MHz:                          5000.0000
CPU min MHz:                          800.0000
BogoMIPS:                             7202.00
Flags:                                fpu ....
Virtualization:                       VT-x
L1d cache:                            384 KiB (8 instances)
L1i cache:                            256 KiB (8 instances)
L2 cache:                             4 MiB (8 instances)
L3 cache:                             16 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-15
Vulnerability Gather data sampling:   Vulnerable: No microcode
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Vulnerable: Clear CPU buffers attempted, no microcode; SMT vulnerable
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==2.1.1
[pip3] pytorch-lightning==2.4.0
[pip3] torch==2.4.1
[pip3] torch_cluster==1.6.3+pt24cu121
[pip3] torchmetrics==0.11.4
[pip3] torchvision==0.19.1
[pip3] triton==3.0.0
[conda] No relevant packages

cc @ezyang @albanD @gqchen @pearu @nikitaved @soulitzer @Varal7 @xmfan @jianyuh @mruberry @walterddr @xwang233 @Lezcano

**Comment by ayghri** [comment]
> The derivation is correct for all I know (cheeky for me to say, as I wrote it). Note that there are a few things at play here:
> 
> 1. The input to `linalg.cholesky` must be symmetric, as per the docs.
>    1.1. The fact that we internally just use the upper or lower triangular part of the matrix is an implementation detail, as this is not specified in the docs.
> 2. As such, computing `linalg.cholesky(x)` is mathematically the same as computing `linalg.cholesky(0.5*(x+x.mT))` given that `x -> 0.5*(x+x.mT)` is the orthogonal projection onto the symmetric matrices (and thus it's the identity on the symmetric matrices). This projection we call `\pi` in the derivation.
> 3. Finally, you since those two functions above are mathematically the same, it is clear that we can compute the derivative of the latter one and will be the same as the derivative of the former one. As such, apply the chain rule to the latter function and you get the formula above.
> 
> Note that in the paper that you link, they assume that the input is triangular. This is why they get a gradient that is "twice as large" as the one in the formula we have implemented.


I got your point and that was also my approach as well. I found this `S + S.T - diag(S)` counter-intuitive at the beginning. You can check [Gadient w.r.t to a symmetric matrix](https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=1101459), it's generic and doesn't apply for lower triangular matrices only.

The catch, I think, is that `0.5*(x+x.mT)` is still symmetric and the reverse grad still needs to be computed under that constraint, and thus the symmetrization from the mentioned paper still holds.

In the example above, if we adopt cholesky based on the lower-half, before symmetrization we get $[[\overline{s}, 0],[\overline{t},\overline{u}]$, using the `0.5(x+x.mT)` that transforms is into $[[\overline{s}, \overline{t}/2],[\overline{t}/2,\overline{u}]$, which is exactly why we see the discrepancy between the current implementation and the grad I got when I manually compute `L`.

**Comment by lezcano** [comment]
I don't have the bandwidht to read that paper, but I am sure that if you use `S + S.mT + diag(S)` is because you are working with triangular matrices, as that operation basically maps a triangular matrix into its associated symmetric matrix. Now, if you apply it to a symmetric matrix, you get all the off-diagonal entries multiplied by two, which wouldn't make any sense.

**Comment by ayghri** [comment]
![proof](https://github.com/user-attachments/assets/1dc0c512-4a6e-44ec-93a3-3a20d3e7b1fd)

This is the shortest proof I could come up with. Please let me know where I got the math wrong.

**Comment by lezcano** [comment]
here, in mathy terms, you are computing the differential and the adjoint to the differential of the function $\phi$. The math is correct, but again, it's unrelated to `linalg.cholesky`.

Another way you can see this is:
- Come up with your favourite way of getting a symmetric matrix (`0.5 * (A + A.mT)`, `A + A.mT - A.diagonal()`...)
- Take a tensor with requires_grad=True, and pass it thorugh this symmetrisation function and into cholesky
- Pass the output and (unconstrained) input to https://pytorch.org/docs/stable/generated/torch.autograd.gradcheck.gradcheck.html

The numerical test should pass.
Then, if you do the same with your formula, either your formula will fail (it's not correct) or will pass (both formulas compute the same quantity).

**Comment by ayghri** [comment]
My bad. I thought your comment was about the changes in `cholesky_backward` not for the comment change in `cholesky_jvp` which is unrelated to the issue at hand. I haven't changed the code in `cholesky_jvp`.
